skbio/diversity/beta/_fast_unifrac.py

```diff
@@ -0,0 +1,444 @@
+                child_index.append((c._leaf_index, c.children[0]._leaf_index,
+                                    c.children[-1]._leaf_index))
+    # handle root, which should be t itself
+    t._leaf_index = curr_index
+    id_index[curr_index] = t
+    # only want to add to the child_index if t has children...
+    if t.children:
+        child_index.append((t._leaf_index, t.children[0]._leaf_index,
+                            t.children[-1]._leaf_index))
+    return id_index, child_index
+
+
+def index_envs(env_counts, tree_index, array_constructor=int):
+    """
+    Calculates the taxon counts in each env (what is an env?)
+
+    Parameters
+    ----------
+    env_counts : np.array
+       Should be the output of count_envs(lines).
+    tree_index : np.array
+       Should be the id_index of index_tree(t).
+    array_constructor: function
+        int by default (may need to change to float later
+        to handle microarray data).
+
+    Returns
+    -------
+    np.array
+        Array of taxon x env with counts of the taxon in each env.
+    """
+    num_nodes = len(tree_index)
+    unique_envs, num_envs = get_unique_envs(env_counts)
+    env_to_index = dict([(e, i) for i, e in enumerate(unique_envs)])
+    result = np.zeros((num_nodes, num_envs), array_constructor)
+    # figure out taxon label to index map
+    node_to_index = {}
+    for i, node in tree_index.items():
+        if node.name is not None:
+            node_to_index[node.name] = i
+    # walk over env_counts, adding correct slots in array
+    for name in env_counts:
+        curr_row_index = node_to_index[name]
+        for env, count in env_counts[name].items():
+            result[curr_row_index, env_to_index[env]] = count
+    # return all the data structures we created; will be useful for other tasks
+    return result, unique_envs, env_to_index, node_to_index
+
+
+def get_branch_lengths(tree_index):
+    """
+    Parameters
+    ----------
+    tree_index: dict
+
+
+    Returns
+    -------
+    np.array
+        Array of branch lengths, in tree index order.
+    """
+    result = np.zeros(len(tree_index), float)
+    for i, node in tree_index.items():
+        try:
+            if node.length is not None:
+                result[i] = node.length
+        except AttributeError:
+            pass
+    return result
+
+
+def _branch_correct(tip_distances, i, j, i_sum, j_sum):
+    """Calculates weighted unifrac branch length correction.
+
+    tip_distances  must be 0 except for tips.
+    """
+    result = tip_distances.ravel()*((i / i_sum)+(j / j_sum))
+    return result.sum()
+
+
+def get_unique_envs(envs):
+    """extract all unique envs from envs dict"""
+    result = set()
+    for v in envs.values():
+        result.update(v.keys())
+    # sort envs for convenience in testing and display
+    return sorted(result), len(result)
+
+
+def bind_to_array(tree_index, a):
+    """Binds tree_index to array a, returning result in list.
```

Comment:
It isn't clear what is going to stay and what is not going to stay as we haven't optimized the implementation to the API so at this point it would be premature